File: python/robot-name/robot_name.py
import random
import string

class Robot:
    used_names = set()

    def __init__(self):
        self._name = None
        self.reset()

    @property
    def name(self):
        return self._name

    def reset(self):
        # Store the current name to avoid generating the same one again
        old_name = self._name
        
        # Clear the current name from the used names set
        if self._name in Robot.used_names:
            Robot.used_names.remove(self._name)
        
        # Generate a new unique name that's different from the old one
        while True:
            # Generate two random uppercase letters
            letters = ''.join(random.choice(string.ascii_uppercase) for _ in range(2))
            
            # Generate three random digits
            digits = ''.join(str(random.randint(0, 9)) for _ in range(3))
            
            # Combine to form the name
            new_name = letters + digits
            
            # Check if the name is unique and different from the old name
            if new_name not in Robot.used_names and new_name != old_name:
                Robot.used_names.add(new_name)
                self._name = new_name
                break

File: javascript/robot-name/robot-name.js
export class Robot {
  static usedNames = new Set();

  constructor() {
    this._name = this.generateUniqueName();
  }

  get name() {
    return this._name;
  }

  reset() {
    const oldName = this._name;
    // Generate a new unique name different from the current one
    do {
      this._name = this.generateUniqueName();
    } while (this._name === oldName);
  }

  generateUniqueName() {
    // Generate a random name that hasn't been used before
    let name;
    do {
      name = this.generateRandomName();
    } while (Robot.usedNames.has(name));

    Robot.usedNames.add(name);
    return name;
  }

  generateRandomName() {
    // Generate 2 random uppercase letters
    const letters = Array(2)
      .fill()
      .map(() => String.fromCharCode(65 + Math.floor(Math.random() * 26)))
      .join('');

    // Generate 3 random digits
    const digits = Array(3)
      .fill()
      .map(() => Math.floor(Math.random() * 10))
      .join('');

    return letters + digits;
  }

  static releaseNames() {
    Robot.usedNames.clear();
  }
}
